[llvm-profdata] Enabled functionality to write split-layout profile

huangjd opened the pull request:
Using the flag `-split_layout` in llvm-profdata merge, the output profile can write profiles with and without inlined function into two different extbinary sections (and their FuncOffsetTable too). The section without inlined functions are marked with `SecFlagFlat` and is skipped by ThinLTO because it provides no useful info.

The split layout feature was already implemented in SampleProfWriter but previously there is no way to use it from llvm-profdata.

huangjd committed:
[llvm-profdata] Enabled functionality to write split-layout profile

Using the flag `-split_layout` in llvm-profdata merge, the output
profile can write profiles with and without inlined function into two
different extbinary sections (and their FuncOffsetTable too). The
section without inlined functions are marked with `SecFlagFlat` and is
skipped by ThinLTO because it provides no useful info.

The split layout feature was already implemented in SampleProfWriter but previously
there is no way from the tool to use it.

ellishg commented on llvm/test/tools/llvm-profdata/sample-split-layout.test:
```diff
@@ -0,0 +1,2 @@
+RUN: llvm-profdata merge --sample --extbinary --split_layout %p/Inputs/sample-profile.proftext -o %t-output
+RUN: diff %t-output %p/Inputs/split-layout.profdata
```
Is there a way to test this without checking in the binary file `split-layout.profdata`?

If you must use a binary file, consider using `printf` to create the file so changes can be reviewed.
https://github.com/llvm/llvm-project/blob/74125fa5ef3e730e53a2685ad7d930a8c685fff9/llvm/test/tools/llvm-cgdata/error.test#L25-L29

minglotus-6 replied:
One way to avoid checking in binary file is to convert it to text format (with `llvm-profdata merge --text`) as input, and convert it back to binary file with a `RUN` line.

WenleiHe replied:
Yes, that sounds good.

huangjd replied:
Note that there is no way to represent flat profile in text mode, I will have to implement that

huangjd replied:
And there's no point to represent that in text mode, because the point of split layout is to allow sample profile reader to skip reading an entire section when used by ThinLTO since samples without inline callsites provide no info to advise inlining. In text mode the profile is read sequentially anyways so you can't skip anything

wlei-llvm replied:
> Note that there is no way to represent flat profile in text mode, I will have to implement that

Maybe we can try `--show-sec-info-only` to check the section size. The size would change if others remove what you adds.

david-xl replied:
Checking section info only sounds good to me.

WenleiHe commented on llvm/tools/llvm-profdata/llvm-profdata.cpp:
```diff
@@ -207,6 +207,12 @@ cl::opt<bool> GenPartialProfile(
     cl::sub(MergeSubcommand),
     cl::desc("Generate a partial profile (only meaningful for -extbinary)"));
+cl::opt<bool> SplitLayout(
+    "split_layout", cl::init(false), cl::Hidden,
```
What's the reason to use `split_layout` as opposed to `split-layout`? You can see all other flags are using `-` in their names.

huangjd replied:
done

WenleiHe commented on llvm/tools/llvm-profdata/llvm-profdata.cpp:
```diff
@@ -207,6 +207,12 @@ cl::opt<bool> GenPartialProfile(
+    "split_layout", cl::init(false), cl::Hidden,
+    cl::sub(MergeSubcommand),
+    cl::desc("Split the profile to two sections with one containing sample "
+             "profiles with inlined functions and the another not (only "
```
the another not -> the other without

huangjd replied:
done

huangjd commented:
> A drive-by comment, what's the use case to write split-layout profile?

ThinLTO uses samples with inlined functions to decide whether to perform inline or not, so samples without inlined functions are useless in this stage, we gather them in one section so that they can be skipped entirely without building a funcoffsettable for them

huangjd committed:
Add CtxSplitLayout support for text mode (for manual inspection only)
Fixed wordings refering to this flag, Previously it is described as
"samples with/out context", which confuses with CSSPGO. Now they are
called "samples/functions with/out inlined functions"

huangjd commented:
Added text mode representation for checking purpose .

wlei-llvm commented on llvm/lib/ProfileData/SampleProfReader.cpp:
```diff
@@ -223,6 +223,11 @@ static bool ParseLine(const StringRef &Input, LineType &LineTy, uint32_t &Depth,
+    // This metadata is only for manual inspection only. We already created a
+    // FunctionSamples and put it in the profile map, so there is no point
+    // to skip profiles even they have no use for ThinLTO.
+    if (Input == StringRef(" !Flat"))
```
I think we need to make sure that profiles in different formats, originating from the same source, to be equivalent. This means that if we do a round-trip conversion—from a "text" profile to an "extbinary" profile and back to a "text" profile—the two text profiles before and after the conversion should be identical. However, it seems that here the "!flat" metadata would be lost because we treat "!Flat" as a no-op, and the extended binary format won't be set to split-layout.

huangjd replied:
Text and Extbinary mode was already not perfectly compatible. There are already various types of info that will be lost when converting Extbinary to Text, for example Profile Symbol List, flags such as PartialProfile, ProfileIsFS etc. Furthermore, the sample profile writer splits flat profiles from non-flat profiles at runtime, and the reader ignores the flag if used by llvm-profdata (instead of used by thinLTO), so the flag is a no-op in this sense, and the round trip output will be identical, unless the user intentionally marked a non-flat function as flat, which doesn't make sense anyways (profile reader could emit a warning in this case but it is also pointless, because 1. in llvm-profdata the writer will correct this, 2. in thinLTO the flat section is skipped intentionally so we don't even know there's non-flat function inside 3. in other compilation mode we read all functions with matching names regardless they are flat or not.)

david-xl replied:
Does losing !flat change the semantics of the profile? If not, they are still equivalent?  On other hand, instead of using meta data, would a comment line be sufficient?

wlei-llvm replied:
Makes sense, thanks for the clarification. I thought for other incompatible cases( Profile Symbol List, PartialProfile, ProfileIsFS, etc), mostly because they don't support text format, but here as the text format is added, so maybe it's not hard to support the compatibility.

WenleiHe commented:
Update change summary: split_layout -> split-layout

> but previously there is no way to use it from llvm-profdata.

How was it previously used?

david-xl reviewed (approved):
LG with Wenlei's last comment addressed.

huangjd commented:
> Update change summary: split_layout -> split-layout
> 
> > but previously there is no way to use it from llvm-profdata.
> 
> How was it previously used?

There is a function in SampleProfileWriter `resetSecLayout` which can be called with the value CtxSplitLayout to change it into split layout. It is not used anywhere in LLVM so there is no user access to it. However in our internal tools we do call this function to create such profile. This patch make this option available to the user.

huangjd committed:
Enforce skipflatprof for text format sample profile

huangjd commented:
I could not reproduce the error on sparcv9 because I don't have the hardware. Those test cases were fine on Linux and I could not reason how would this patch affect them. Could someone also take a look?

rorth commented:
You do, actually: the GCC Compile Farm contains a Solaris/sparcv9 system perfectly equipped to build and test LLVM.

That said, I've done some investigation myself: even in exactly the same configuration as the buildbot (`clang-17` as host compiler, both 1-stage and 2-stage builds) I couldn't reproduce the failures.  I have to conclude that it's somehow related to the fact that the bot is doing incremental builds most of the time, while my builds were always from scratch.

I've now switched the bot to use `clang-19` (19.1.0-rc3) as host compiler and after a fresh build with that one, the failure is gone.  Very strange.
